pkg/datastore/firestore/iterator.go

```diff
@@ -33,10 +36,23 @@ func (it *Iterator) Next(dst interface{}) error {
 		}
 		return err
 	}
+
+	// Update last iterated item as last read doc.
+	it.last = doc
+
 	return doc.DataTo(dst)
 }
 
 func (it *Iterator) Cursor() (string, error) {
-	// Note: Cursor function is not needed in Cloud Firestore.
-	return "", datastore.ErrUnimplemented
+	if it.last == nil {
+		return "", datastore.ErrInvalidCursor
+	}
+
+	cursorVal, err := json.Marshal(it.last)
```

Comment:
So we encoding the whole object to be the cursor. How to access a deep field from the `makeCursorValues` function, for example, the sync status of the application model?

Reply:
I updated the caller side logic already 👀 
https://github.com/pipe-cd/pipe/pull/1864/files#diff-dd8df471b287825d31c09f115e2987964d6937facd2133440d1f703544639771R248-R251
```go
	// Decode last object of previous page stored as opts.Cursor in json format.
	obj := make(map[string]interface{})
	if err := json.Unmarshal([]byte(opts.Cursor), &obj); err != nil {
		return nil, err
	}
```
After the above lines, we should be able to access the value of the cursor by JSON value access operator such as (in your concern): `obj["sync_state"]["status"]`

Reply:
Encoding the whole object can be easy at first but it requires some stuff like encrypting the cursor before sending it to the client (web, pipectl). So how about the cursor should keep only the ordering fields. I mean it should be `map[string]interface{}` as it is but not containing the whole object data, just including the specified fields.

Reply:
I get your point, but building by specified fields is difficult than it looks like 😢 We have to know ( in the encoding side - `iterator.Cursor()` ), whether the current `it.last` object contains "specified fields" or not, that cost more than encrypting thing I guess. We also need to update the "specified fields" list when we're going to support new ordering fields, which increase the maintenance cost too 🤔

Reply:
I will take a deeper look at this later, but at this time I think we can pass the name list of the specified fields to the Iterator in the `Find` then later we still encode the whole object into JSON and use the specified field names to simplify the content of cursor.

Reply:
Nice idea 👍 Addressed on https://github.com/pipe-cd/pipe/pull/1864/commits/9653d7e22de54cbf95eae18043854e681fad128e 🙏